File: strategy-game-frontend/src/core/models/Color.ts
export interface Color {
    red: number,
    green: number,
    blue: number
}

export namespace Color {

    export const BLACK: Color = {
        red: 0,
        green: 0,
        blue: 0
    };

    export const INVALID: Color = {
        red: -255,
        green: -255,
        blue: -255
    };

    export function colorToRgbArray(color: Color): number[] {
        return [color.red / 255, color.blue / 255, color.green / 255];
    }

    export function colorToRgbaArray(color: Color, alpha: number): number[] {
        return [color.red / 255, color.blue / 255, color.green / 255, alpha];
    }

}

File: backend/strategy-game-common-data/src/main/kotlin/io/github/smiley4/strategygame/backend/commondata/utils/Color.kt
package io.github.smiley4.strategygame.backend.commondata.utils

sealed interface Color {

    val redByte: UByte
    val redFloat: Float

    val greenByte: UByte
    val greenFloat: Float

    val blueByte: UByte
    val blueFloat: Float

    fun toRgbByte(): RgbByteColor

    companion object {

        fun rgb(red: UByte, green: UByte, blue: UByte) = RgbByteColor(red, green, blue)

        val BLACK = rgb(0u, 0u, 0u)

    }

}

class RgbByteColor(
    val red: UByte,
    val green: UByte,
    val blue: UByte
) : Color {

    override val redByte: UByte
        get() = red

    override val redFloat: Float
        get() = red.toFloat() / 255.0f

    override val greenByte: UByte
        get() = green

    override val greenFloat: Float
        get() = green.toFloat() / 255.0f

    override val blueByte: UByte
        get() = blue

    override val blueFloat: Float
        get() = blue.toFloat() / 255.0f

    override fun toRgbByte() = this
}